omp/reorder/rcm_kernels.cpp

```diff
@@ -0,0 +1,860 @@
+            head += n;
+
+            omp_unset_lock(&read_lock);
+            return std::make_pair(data + old_head, n);
+        } else {
+            // This "spinlock" is fine without cmpxchg, exclusive waiter-access
+            // is guaranteed by the outer lock.
+            while (true) {
+                int val_threads_working;
+#pragma omp atomic read
+                val_threads_working = *threads_working;
+
+                if (val_threads_working == 0) {
+                    break;
+                }
+
+                // No measureable effect on performance.
+                GKO_MM_PAUSE();
+            }
+
+            // Recalculate the chunk size, now that all threads are finished.
+            const auto n = chunk_size();
+            if (n > 0) {
+#pragma omp atomic update
+                ++(*threads_working);
+
+                IndexType old_head = head;
+                head += n;
+
+                omp_unset_lock(&read_lock);
+                return std::make_pair(data + old_head, n);
+            } else {
+                omp_unset_lock(&read_lock);
+                return std::make_pair(nullptr, 0);
+            }
+        }
+    }
+};
+
+
+template <typename IndexType>
+inline void reduce_neighbours_levels(const IndexType num_vertices,
+                                     const IndexType *const row_ptrs,
+                                     const IndexType *const col_idxs,
+                                     IndexType *const local_to_insert,
+                                     size_type *const local_to_insert_size,
+                                     IndexType *const levels,
+                                     const IndexType node)
+{
+    IndexType level;
+#pragma omp atomic read
+    level = levels[node];
+
+    const auto new_neighbour_level = level + 1;
+
+    // Go through neighbours.
+    const auto row_start = row_ptrs[node];
+    const auto row_end = row_ptrs[node + 1];
+
+    for (auto neighbour_i = row_start; neighbour_i < row_end; ++neighbour_i) {
+        const auto neighbour = col_idxs[neighbour_i];
+
+        // Immediately evaluated lambda expression,
+        // to enable continuing the outer loop.
+        const auto reduce_level_atomic = [&]() {
+            // Atomically check if we have lower new level,
+            // then set, implemented with cmpxchg.
+
+            IndexType old_neighbour_level;
+            do {
+#pragma omp atomic read
+                old_neighbour_level = levels[neighbour];
+
+                const auto is_new_smaller =
+                    new_neighbour_level < old_neighbour_level;
+
+                if (!is_new_smaller) {
+                    return true;
+                }
+
+            } while (!__atomic_compare_exchange_n(
+                &levels[neighbour], &old_neighbour_level, new_neighbour_level,
+                true, std::memory_order::memory_order_acq_rel,
+                std::memory_order::memory_order_acquire));
```

Comment:
[The previous discussion](https://github.com/ginkgo-project/ginkgo/pull/500#discussion_r479238475) on this topic provided a decent introduction to the downsides. Possible solutions would be `reinterpret_cast<std::atomic<...>*>` (UB), type punning using unions (UB, but legal in GC) or just leaving it as-is, since all compilers we use seem to support this intrinsic (Didn't check AppleClang and Windows in detail, though!)

Reply:
See my answer to the exact same complaint in the review above. I checked support for all compilers and targets, indeed MSVC calls its atomic builtin differently. I already planned on introducing a define here.